ASAA-39 + ASAA-40 - Feature Toggles from JSON File, Setup Core Feature Interfaces

LuisRocketRangel opened the pull request:
**ASAA-39:**
Requirements:
-Add Feature Toggle models
-Add ability to read Feature Toggles from Json files -Added a couple of Unit Tests
-Adding a PR now as I'm being rolled off Bench soon, please call me out as needed lol
Solution:
-Did above

**ASAA-40:**
Requirements:
-Build core feature toggle interface and supporting classes
Solution:
-Did above

LuisRocketRangel committed:
ASAA-39-Feature Toggles from Json File

Requirements:
-Add Feature Toggle models
-Add ability to read Feature Toggles from Json files
-Added a couple of Unit Tests

Solution:
-Did above

jaredstockton commented on domain/src/main/java/com/bottlerocketstudios/brarchitecture/domain/repositories/FeatureToggleRepository.kt:
```diff
@@ -0,0 +1,11 @@
+import kotlinx.coroutines.flow.Flow
+
+@Suppress("TooManyFunctions")
+interface FeatureToggleRepository : com.bottlerocketstudios.brarchitecture.domain.models.Repository {
```
@LuisRocketRangel happy to sync more on the overarching approach for the feature toggles. Note that I did add some sample interfaces/classes/implementations to AASA-40 takes a different approach than shown here :)

cc: @BottleRocket-Colin

jaredstockton commented on domain/src/main/java/com/bottlerocketstudios/brarchitecture/domain/models/FeatureToggle.kt:
```diff
@@ -0,0 +1,8 @@
+package com.bottlerocketstudios.brarchitecture.domain.models
+
+data class FeatureToggle(
```
@LuisRocketRangel Having a sealed class represent the feature toggles (as seen in AASA-40) would be ideal!

jaredstockton replied:
That being said, in general it is preferable to use non-nullable types in domain models where able, especially for Collection types (List, Set, etc), Strings, and Booleans (unless you need the absence of a Boolean value, denoted by null)

jaredstockton commented on app/src/main/assets/featureToggles.json:
```diff
@@ -0,0 +1,14 @@
+[
```
@LuisRocketRangel You could also add the json inside a kotlin file and follow https://github.com/BottleRocketStudios/Android-ArchitectureDemo/blob/develop/docs/BEST_PRACTICES.md#language-injections even though this is a valid approach as well.

LuisRocketRangel replied:
I better follow the Best Practices, sorry about that! It's fixed!

jaredstockton commented on domain/src/main/java/com/bottlerocketstudios/brarchitecture/domain/models/FeatureToggle.kt:
```diff
@@ -0,0 +1,8 @@
+data class FeatureToggle(
+    val name: String?,
+    val value: Boolean?,
+    val defaultValue: Boolean?,
```
@LuisRocketRangel This might come in a future PR but we'll definitely want to persist the feature toggles locally (likely via DataStore) so that the value persists across app restarts and is changeable via some Dev Options style UI

LuisRocketRangel committed:
ASAA-39-Jared Feedback Part I

Requirements:
-Change JSON Asset file to JSON String

Solution:
-Did above

jacob-taggart committed:
ASAA-39, ASAA-40 - change approach to FeatureToggle to use sealed class, add different type of DTO, edit JSON, add tests

jacob-taggart committed:
ASAA-39, ASAA-40 - remove unnecessary call that was used for dev testing
